### DISCUSSION TITLE
How to upload application program with nRF ToolBox

### DISCUSSION THREAD

**Discussion by devgnrb** [kaidyth/nrf52_bootloader#20] [Q&A]
Hi, 

I've burn my Arduino Nano 33 BLE Bootloader with this bootloader and it seems to be working (Flashing led red/blue and detecting it in Bluetooth). 

Here is my simple program to measure digital pin 2: 

```
void setup() {
  pinMode(2, OUTPUT);
}

void loop() {
    digitalWrite(2, LOW);
    delay(500);
    digitalWrite(2, HIGH);

}
```

I've created an HEX file of this program with Arduino IDE and doing this to create a zip: 
`nrfutil.exe pkg generate --application .\spineo_arduino.ino.hex --key-file .\private.pem --hw-version 52 --sd-req 0xCA --application-version 1 app.zip`

I'm able to upload this .zip file thanks to nRF Toolbox without any issues but even this simple program doesn't work. After having upload it in the arduino, leds are ON instead of being of and nothing appens... Do anyone have an idea why it doesn't do what I want ?

Thanks

**Comment by charlesportwoodii** [comment]
I really don't use Arduino or support Arduino code on this bootloader but I can try to help. Candidly though if you're looking to use Arduino code you'd be better off using mbed instead. The Arduino BlueFruit library won't ever load on this bootloader due to an upstream issue with Arduino. This bootloader is really meant for native C code using the NRF SDK.

Can you share more information about your setup? Pin 2 is D2 on the board itself, not the built in LEDs.

Do you have a JLink device, or multimeter you can connect to Pin D2 and see if it's changing state? I'd have to find my Nano 33 BLE from storage to try to debug separately. The DEBUG board variant of the bootloader can help with debugging. If the application is crashing you'd be able to see an application crash via JLink.

**Comment by devgnrb** [comment]
All right, I'll try mbed and see if it changes something... 
Yes I know, I've a multimeter on D2 so I can see if the board is running. I've tried with LED_PWR but it didn't worked so I tried with this method. 

I'm flashing bootloader with J-Link EDU Mini but I don't know yet (and I didn't think about it) how to be in debug mode but I'll look for it.

**Reply by devgnrb** [reply]
I've tried with mbed studio to blink the builtin led, compile it and push the .hex file with nrf toolbox: 
```
/* mbed Microcontroller Library
 * Copyright (c) 2019 ARM Limited
 * SPDX-License-Identifier: Apache-2.0
 */

#include "mbed.h"

// Blinking rate in milliseconds
#define BLINKING_RATE     500ms


int main()
{
    // Initialise the digital pin LED1 as an output
    DigitalOut led(P0_13);

    while (true) {
        led = !led;
        ThisThread::sleep_for(BLINKING_RATE);
    }
}
```
But unfortynatly, it's still doing the same... I have the power led and the builtin led ON, I don't know why and I can't control anything...

**Comment by charlesportwoodii** [comment]
Sorry for any confusion. I meant the Arduino mbed bootloader, not mbed code itself. As I mentioned earlier this really isn't meant for Arduino code - it's purpose is really for native-c code instead.

I would be willing to bet that the default linker settings in Arduino Studio still haven't been updated for Nordic Soft Device 7.

If there's a RAM conflict, you can verify it by doing the following:
1. Build/Install the DEBUG variant for your board.
2. Attach a JLink and see what it's outputting. JLink reports the memory error - I don't remember the exact error message but it's obvious once you see the logs.

The Application RAM regions need to be appropriately aligned, and if they're not software will panic. If I recall, the default Arduino RAM regions are setup inappropriately for use with SD7 - since all of the chips are still running and using SD6. You'll need to find the linker (ld) file Arduino Studio is using and adjust the regions there.

This bootloader uses SD7 (but if you want an SD6 bootloader, I think version 0.4.0 uses that.

For SD7, it'll need to be as follows:

```
FLASH (rx) : ORIGIN = 0x27000, LENGTH = 0xc9000 
RAM (rwx) :  ORIGIN = 0x20002280, LENGTH = 0x3dd80
uicr_bootloader_start_address (r) : ORIGIN = 0x00000FF8, LENGTH = 0x4
```

Someone else had a similar problem with Ardunio Studio on a custom board and this was the fix for it as well - ref: https://github.com/kaidyth/nrf52_bootloader/issues/15#issuecomment-651410953

**Comment by devgnrb** [comment]
All right my bad... 
So, I'm trying with SD6 and version 0.4.0 with debug enabled. After having installed bootloader and app via Bluetooth, I'm having these logs: 
```
<info> app: Kaidyth DFU: Inside main

<debug> app: Kaidyth DFU: Timers initialized

<debug> app: Kaidyth DFU: BSP initialized

<debug> app: In nrf_bootloader_init

<debug> nrf_dfu_settings: Calling nrf_dfu_settings_init()...

<debug> nrf_dfu_flash: Initializing nrf_fstorage_nvmc backend.

<debug> nrf_dfu_settings: Using settings page.

<debug> nrf_dfu_settings: Copying forbidden parts from backup page.

<debug> nrf_dfu_settings: Destination settings are identical to source, write not needed. Skipping.

<info> nrf_dfu_settings: Backing up settings page to address 0xFE000.

<debug> nrf_dfu_settings: Destination settings are identical to source, write not needed. Skipping.

<debug> app: Enter nrf_bootloader_fw_activate

<info> app: No firmware to activate.

<debug> app: App is valid

<info> nrf_dfu_settings: Backing up settings page to address 0xFE000.

<debug> nrf_dfu_settings: Destination settings are identical to source, write not needed. Skipping.

001000

<debug> app: Disabling interrupts. NVIC->ICER[0]: 0x100000

```

Is there an obvious problem I don't see ?

**Comment by charlesportwoodii** [comment]
I'm sorry you're having issues. If the app is failing to launch then you still have linker issues. Back when I had a Nano 33 BLE it was still using Arduino NRF52 Core. If you're using the 0.4.0 board please ensure your app is packaged for the SD6 soft device if you're using 0.4.0: 0xB6 rather than the SD7. In the year since things might have changed.

My recommendation for you would be to either:

1. Write your test bench in native C with the linker settings I provided.
2. Use another bootloader.

This bootloader is not designed, nor intended to support Arduino or Mbed. It's sole purpose is to enhance the existing Nordic Secure Soft Device with a few of the Adafruit Bootloader feature sets. I use them exclusively for running custom C software and testing specific firmware features on commercial boards. If you're not working with the NRF SDK directly, I strongly recommend using either the bootloader that came with your board, or the Adafruit one, since it supports other Soft Device features this one cannot due to how the Nordic Secure Bootloader is implemented.

I don't use Arduino at all. I can confirm that the following Arduino loop works on my nano 33 ble, with SD7, using the linker settings I provided on the latest bootloader version.

```
void setup() { Serial.begin(115200); }
void loop()
{
  Serial.println("loop");
  delay(1000);
}
```

However beyond that any issue is an application issue you'll need to address independently. I can't provide support for Arduino code.
